coreax/metrics.py

```diff
@@ -95,86 +91,104 @@ def __init__(self, kernel: coreax.kernel.Kernel, precision_threshold: float = 1e
+        both datasets and reference weights are not uniform in the weighted case, the
+        calculation is done block-wise to limit memory requirements.
 
-        :param x: The original :math:`n \times d` data
-        :param y: :math:`m \times d` dataset
+        :param reference_data: An instance of the class :class:`coreax.data.Data`
+        :param comparison_data: An instance of the class :class:`coreax.data.Data` to
+            compare against ``reference_data``
         :param block_size: Size of matrix block to process, or :data:`None` to not split
             into blocks
-        :param weights_x: An :math:`1 \times n` array of weights for associated points
-            in ``x``, or :data:`None` if not required
-        :param weights_y: An :math:`1 \times m` array of weights for associated points
-            in ``y``, or :data:`None` if not required
         :return: Maximum mean discrepancy as a 0-dimensional array
         """
         # Format inputs
-        x = jnp.atleast_2d(x)
-        y = jnp.atleast_2d(y)
-        if weights_x is not None:
-            weights_x = jnp.atleast_1d(weights_x)
-        if weights_y is not None:
-            weights_y = jnp.atleast_1d(weights_y)
-
-        num_points_x = len(x)
-        num_points_y = len(y)
-
-        if weights_y is None:
-            if block_size is None or block_size > max(num_points_x, num_points_y):
-                return self.maximum_mean_discrepancy(x, y)
-            return self.maximum_mean_discrepancy_block(x, y, block_size)
+        reference_points = jnp.atleast_2d(reference_data.data)
+        num_reference_points = len(reference_points)
+        reference_weights = jnp.atleast_1d(reference_data.weights)
+        if jnp.allclose(
+            reference_weights,
+            jnp.broadcast_to(1 / num_reference_points, (num_reference_points,)),
+        ):
+            reference_weights = None
+
+        comparison_points = jnp.atleast_2d(comparison_data.data)
+        num_comparison_points = len(comparison_points)
+        comparison_weights = jnp.atleast_1d(comparison_data.weights)
+        if jnp.allclose(
+            comparison_weights,
+            jnp.broadcast_to(1 / num_comparison_points, (num_comparison_points,)),
+        ):
+            comparison_weights = None
+
+        if comparison_weights is None:
+            if block_size is None or block_size > max(
+                num_reference_points, num_comparison_points
+            ):
+                return self.maximum_mean_discrepancy(
+                    reference_points, comparison_points
+                )
+            return self.maximum_mean_discrepancy_block(
+                reference_points, comparison_points, block_size
+            )
 
         if (
             block_size is None
-            or weights_x is None
-            or block_size > max(num_points_x, num_points_y)
+            or reference_weights is None
```

Comment:
I recognise this is a pre-existing issue, but if `block_size is None` and `reference_weights is not None` then `reference_weights` is not used. If this is intended behaviour, it should at least warn the user!

Reply:
Yes this seems wrong, Tom C is working a sweeping change for this

Reply:
I have changed this behaviour slightly but will be addressed more fully by #613